File: leetcode/Math/171. Excel Sheet Column Number.cpp
// Author: Shengjia Yan
// Date: 2018-05-29 Tuesday
// Email: i@yanshengjia.com
// Time Complexity: O(n)  n == length of s
// Space Complexity: O(1)


class Solution {
public:
    int titleToNumber(string s) {
        int length = s.size();
        int res = 0;
        
        for (int i=length-1; i>=0; i--) {
            int exp = length - i - 1;
            res += pow(26, exp) * (s[i] - 64);
        }

        return res;
    }
};

File: leetcode/Math/171. Excel Sheet Column Number.py
"""
Given a column title as appear in an Excel sheet, return its corresponding column number.

For example:

    A -> 1
    B -> 2
    C -> 3
    ...
    Z -> 26
    AA -> 27
    AB -> 28 
    ...
Example 1:

Input: "A"
Output: 1
Example 2:

Input: "AB"
Output: 28
Example 3:

Input: "ZY"
Output: 701


Solution:
26-coded Convention
"""


# Time: O(n), n is the length of s
# Space: O(1)
class Solution:
    def titleToNumber(self, s: str) -> int:
        res = 0
        l = len(s)
        for i in range(l):
            power = l - i - 1
            res += 26 ** power * (ord(s[i]) - ord('A') + 1)
        return res
        
